bind/contrib/mdnkit-2.2.3-src/tools/mdnconv/mdnconv.1
.\" $Id: mdnconv.1,v 1.1.1.1 2002/11/18 22:25:28 bbraun Exp $
.\"
.\" Copyright (c) 2000,2001 Japan Network Information Center.
.\" All rights reserved.
.\"  
.\" By using this file, you agree to the terms and conditions set forth bellow.
.\" 
.\" 			LICENSE TERMS AND CONDITIONS 
.\" 
.\" The following License Terms and Conditions apply, unless a different
.\" license is obtained from Japan Network Information Center ("JPNIC"),
.\" a Japanese association, Kokusai-Kougyou-Kanda Bldg 6F, 2-3-4 Uchi-Kanda,
.\" Chiyoda-ku, Tokyo 101-0047, Japan.
.\" 
.\" 1. Use, Modification and Redistribution (including distribution of any
.\"    modified or derived work) in source and/or binary forms is permitted
.\"    under this License Terms and Conditions.
.\" 
.\" 2. Redistribution of source code must retain the copyright notices as they
.\"    appear in each source code file, this License Terms and Conditions.
.\" 
.\" 3. Redistribution in binary form must reproduce the Copyright Notice,
.\"    this License Terms and Conditions, in the documentation and/or other
.\"    materials provided with the distribution.  For the purposes of binary
.\"    distribution the "Copyright Notice" refers to the following language:
.\"    "Copyright (c) Japan Network Information Center.  All rights reserved."
.\" 
.\" 4. Neither the name of JPNIC may be used to endorse or promote products
.\"    derived from this Software without specific prior written approval of
.\"    JPNIC.
.\" 
.\" 5. Disclaimer/Limitation of Liability: THIS SOFTWARE IS PROVIDED BY JPNIC
.\"    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\"    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
.\"    PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL JPNIC BE LIABLE
.\"    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\"    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\"    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\"    BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\"    WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\"    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\"    ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
.\" 
.\" 6. Indemnification by Licensee
.\"    Any person or entities using and/or redistributing this Software under
.\"    this License Terms and Conditions shall defend indemnify and hold
.\"    harmless JPNIC from and against any and all judgements damages,
.\"    expenses, settlement liabilities, cost and other liabilities of any
.\"    kind as a result of use and redistribution of this Software or any
.\"    claim, suite, action, litigation or proceeding by any third party
.\"    arising out of or relates to this License Terms and Conditions.
.\" 
.\" 7. Governing Law, Jurisdiction and Venue
.\"    This License Terms and Conditions shall be governed by and and
.\"    construed in accordance with the law of Japan. Any person or entities
.\"    using and/or redistributing this Software under this License Terms and
.\"    Conditions hereby agrees and consent to the personal and exclusive
.\"    jurisdiction and venue of Tokyo District Court of Japan.
.\"
.TH MDNCONV 1 "Mar 3, 2001"
.\"
.SH NAME
mdnconv \- codeset converter for named.conf and zone master files
.\"
.SH SYNOPSIS
\fBmdnconv\fP [\fIoptions..\fP] [\fIfile\fP...]
.\"
.SH DESCRIPTION
.B mdnconv
is a codeset converter for named configuration files and zone master files.
.B mdnconv
performs codeset conversion specified either by the command-line arguments
or by the configuration file,
and writes the converted text to stdout.
.PP
If file name is specified,
.B mdnconv
converts the contents of the file.  Otherwise,
.B mdnconv
converts
.IR stdin .
.PP
Since
.B mdnconv
is specifically designed for converting internatinalized domain names,
it may not be suitable as a general codeset converter.
.\"
.SH "OPERATION MODES"
.B mdnconv
has two operation modes.
.PP
One is a mode to convert local-encoded domain names to IDN-encoded
one.  Usually this mode is used for preparing domain names to be
listed in named configuration files or zone master files.
In this mode, the following processes are performed in addition to
the codeset (encoding) conversion.
.RS 2
.IP \- 2
local mapping
.IP \- 2
standard domain name preperation (NAMEPREP)
.RE
.PP
Furthermore, delimiter mapping is also performed in case \-delimitermap
command line option is specified.
.PP
The other mode is a reverse conversion, from IDN-encoded domain name to
local-encoded domain names.
In this mode, local mapping and NAMEPREP are not performed since
IDN-encoded names should already be normalized.
Instead, a check is done in order to make sure the IDN-encoded domain name
is properly NAMEPREP'ed.  If it is not, the name will be output in
IDN encoding, not in the local encoding.
.\"
.SH OPTIONS
Normally
.B mdnconv
reads system's default configuration file (mdn.conf) and perform
conversion or name preparation according to the parameters specified in
the file.  You can override the setting in the configuration file by
various command line options below.
.TP 4
\fB\-in\fP \fIin-code\fP, \fB\-i\fP \fIin-code\fP
Specify the codeset name of the input text.
Any of the following codeset names can be specified.
.RS 4
.IP "\(bu" 2
Any codeset names which \fIiconv_open()\fP library function accepts
.IP "\(bu" 2
\f(CWAMC-ACE-Z\fR
.IP "\(bu" 2
\f(CWRACE\fR
.IP "\(bu" 2
\f(CWDUDE\fR
.IP "\(bu" 2
\f(CWUTF-8\fR
.IP "\(bu" 2
Any alias names for the above, defined by the codeset alias file.
.RE
.IP "" 4
If this option is not specified, the default codeset is determined
from the locale in normal conversion mode.
In reverse conversion mode, the default codeset is the IDN encoding
specified by the configuration file (``idn-encoding'' entry).
.TP 4
\fB\-out\fP \fIout-code\fP, \fB\-o\fP \fIout-code\fP
Specify the codeset name of the output text. \fIout-code\fP can be any
codeset name that can be specified for 
.B \-in
option.
.IP "" 4
If this option is not specified, the default is the IDN encoding
specified by the configuration file (``idn-encoding'' entry) in
normal conversion mode.
In reverse conversion mode, the default codeset is determined from
the locale.
.TP 4
\fB\-conf\fP \fIpath\fP, \fB\-c\fP \fIpath\fP
Specify the pathname of mDNkit configuration file (``mdn.conf'').
If not specified, system's default file is used, unless \-noconf
option is specified.
.TP 4
\fB\-noconf\fP, \fB\-C\fP
Specify that no configuration file is to be used.
.TP 4
\fB\-reverse\fP, \fB\-r\fP
Specify reverse conversion mode.
.br
If this option is not specified, the normal conversion mode is used.
.TP 4
\fB\-nameprep\fR \fIversion\fR, \fB\-n\fR \fIversion\fR
Specify the version of NAMEPREP.
The following is a list of currently available
.IR version s.
.RS 4
.IP \f(CWnameprep-03\fR 4
Perform NAMEPREP according to the Internet Draft
``draft-ietf-idn-nameprep-03.txt'' (NAMEPREP-03).
.IP \f(CWnameprep-05\fR 4
Perform NAMEPREP according to the Internet Draft
``draft-ietf-idn-nameprep-05.txt'' (NAMEPREP-05).
.IP \f(CWnameprep-06\fR 4
Perform NAMEPREP according to the Internet Draft
``draft-ietf-idn-nameprep-06.txt'' (NAMEPREP-06).
.RE
.IP "" 4
This option is a shorthand for specifying the 4 options below.
.PP
.RS 8
.ft CW
-map 
.I version
.br
-normalize
.I version
.br
-prohibit
.I version
.br
-unassigned
.I version
.ft R
.RE
.TP 4
\fB\-nonameprep\fR, \fB\-N\fR
Specify to skip NAMEPREP process (or NAMEPREP verification process
in the reverse conversion mode).
.TP 4
\fB\-map\fR \fImap\fR
Specify the name of the mapping rule for NAMEPREP mapping process.
Currently, following
.IR map s
are available.
.RS 4
.IP \f(CWnameprep-03\fR 4
Use the list of mappings specified by NAMEPREP-03 draft.
.IP \f(CWnameprep-05\fR 4
Use the list of mappings specified by NAMEPREP-05 draft.
.IP \f(CWnameprep-06\fR 4
Use the list of mappings specified by NAMEPREP-06 draft.
.IP \f(CWfilemap:\fR\fIpath\fR 4
Use list of mappings specified by mapfile
.IR path .
See mdn.conf(5) for the format of a mapfile.
.RE
.IP "" 4
This option can be specified more than once.
In that case, each mapping will be performed in the order of the
specification.
.br
If neither this option nor ``\-nameprep'' are specified,
the schemes specified in the configuration file (``nameprep-map'' entry)
are used.
.TP 4
\fB\-normalize\fP \fIscheme\fP
Specify the normalization scheme for NAMEPREP normalization process.
Currently, following \fIscheme\fPs are available:
.RS 4
.IP \f(CWunicode-form-kc\fR
Unicode Normlization Form KC, also described in
``Unicode Technical Report #15: Unicode Normalization Forms'',
available from unicode.org.
.IP \f(CWunicode-form-kc/3.0.1\fR
Unicode Normlization Form KC, version 3.0.1
.IP \f(CWunicode-form-kc/3.1.0\fR
Unicode Normlization Form KC, version 3.1.0
.IP \f(CWnameprep-03\fR
Synonim for ``unicode-form-kc/3.0.1''.
.IP \f(CWnameprep-05\fR
Synonim for ``unicode-form-kc/3.1.0''.
.IP \f(CWnameprep-06\fR
Synonim for ``unicode-form-kc/3.1.0''.
.RE
.IP "" 4
This option can be specified more than once.
In that case, each normalization will be performed in the order of the
specification.
.br
If neither this option nor ``\-nameprep'' are specified,
the schemes specified in the configuration file (``nameprep-normalize'' entry)
are used.
.TP 4
\fB\-prohibit\fR \fIset\fR
Specify the set of prohibited characters.
Currently, following
.IR set s
are available.
.RS 4
.IP \f(CWnameprep-03\fR 4
Use list of prohibit characters specified by NAMEPREP-03 draft.
.IP \f(CWnameprep-05\fR 4
Use list of prohibit characters specified by NAMEPREP-05 draft.
.IP \f(CWnameprep-06\fR 4
Use list of prohibit characters specified by NAMEPREP-06 draft.
.IP \f(CWfileset:\fR\fIpath\fR 4
Use list of characters specified by setfile
.IR path .
See mdn.conf(5) for the format of a setfile.
.RE
.IP "" 4
This option can be specified more than once.  In this case, the set of
prohibited characters is the union of the specified sets.
.br
If neither this option nor ``\-nameprep'' is specified,
list of characters specified by the configuration file
(``nameprep-prohibit'' entry) is used.
.TP 4
\fB\-unassigned\fR \fIset\fR
Specify the set of unassigned codepoints.
Currently, following
.IR set s
are available.
.RS 4
.IP \f(CWnameprep-03\fR 4
Use the list of unassigned characters specified by NAMEPREP-03 draft.
.IP \f(CWnameprep-05\fR 4
Use the list of unassigned characters specified by NAMEPREP-05 draft.
.IP \f(CWnameprep-06\fR 4
Use the list of unassigned characters specified by NAMEPREP-06 draft.
.IP \f(CWfileset:\fR\fIpath\fR 4
Use list of characters specified by setfile
.IR path .
See mdn.conf(5) for the format of a setfile.
.RE
.IP "" 4
Note that unassigned codepoint checking is not performed
unless \-unassigncheck option is specified.
.br
This option can be specified more than once.  In this case, the set of
prohibited characters is the union of the specified sets.
.br
If neither this option nor ``\-nameprep'' is specified,
list of characters specified by the configuration file
(``nameprep-unassigned'' entry) is used.
.TP 4
\fB\-unassigncheck\fR, \fB\-u\fR
Specify unassigned codepoint check should be performed.
Unassigned codepoint check is disabled by default.
.TP 4
\fB\-delimiter\fR \fIcodepoint\fP
Specify the character to be mapped to domain name delimiter (period).
This option can be specified more than once in order to specify multiple
characters.
.br
If this option is not specified, characters specified by the
configuration file (``delimiter-map'') is used.
.br
Note that delimiter mapping is not performed unless \-delimitermap option
is specified.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-delimitermap\fR, \fB\-d\fR
Specify delimiter mapping should be performed.
By default, delimiter mapping is disabled.
.br
This option is only meaningful in the normal conversion mode.
.TP 4
\fB\-whole\fP, \fB\-w\fP
Perform local mapping, nameprep and conversion to output codeset for the entire
input text.  If this option is not specified, only non-ASCII characters
and their surrounding texts will be processed.
See ``CONVERSION MECHANISM'' for details.
.TP 4
\fB\-alias\fP \fIpath\fP, \fB\-a\fP \fIpath\fP
Specify a codeset alias file.  It is a simple text file, where
each line has a pair of alias name and real name separated by one
or more white spaces like below:
.nf
.ft CW

    \fIalias-codeset-name\fP    \fIreal-codeset-name\fP

.ft R
.fi
Lines starting with ``#'' are treated as comments.
.TP 4
\fB\-flush\fP
Force line-buffering mode.
.TP 4
\fB\-version\fP, \fB\-v\fP
Print version information and quit.
.\"
.SH LOCAL CODESET
mdnconv guesses local codeset from locale and environment variables.
See the ``LOCAL CODESET'' section in mdn.conf(5) for more details.
.\"
.SH CONVERSION MECHANISM
Here is a brief step-by-step description of how
.B mdnconv
does its job.
.IP "1. input-side convnersion" 4
.B mdnconv
reads input text line by line, and converts them from the input codeset
to UTF-8 encoding.  Basically this conversion is done using iconv() utility,
but for some special codesets like RACE, special converter will be used.
.br
Normally the whole line is converted, but if the input codeset is an
ACE (ASCII-compatible encoding, such as RACE), a special conversion
prodedure shown below is used:
.RS 4
.IP 1) 3
Find substrings which are valid ASCII domain names (i.e. constructed from
alphanumeric characters and hyphens) in the line.
.IP 2) 3
For each substring found, try converting it to UTF-8.
.IP 3) 3
If the conversion succeeds, replace the substring with the conversion
result. Otherwise, the original substring is retained.
.RE
.IP "2. finding where to process" 4
Then
.B mdnconv
looks for (part of) internationalized domain name in the converted line,
and picks them up for further processing.
.br
.B mdnconv
recognizes the character sequence having the following property as
an internationalized domain name.
.RS 4
.IP "\(bu" 2
containing at least one non-ASCII character, and
.IP "\(bu" 2
consisting of legal domain name characters (alphabets, digits, hypens),
non-ASCII characters and period.
.RE
.IP "" 4
If you are not quite sure what this means (as I), 
the following regular expression may help you.
.nf
.ft CW

    [-A-Za-z0-9.]*<non-ASCII char>([-A-Za-z0-9.]|<non-ASCII char>)*

.ft P
.fi
For example, only ``www.###.com'' will be picked up from
the following lines (`#' denotes non-ASCII character).
.nf
.ft CW

    www.###.com.    IN CNAME serv.xyz.com.
    serv.xyz.com.   IN A     10.0.0.1

.ft P
.fi
If \-whole option is specified, this step is skipped and the entire line
is subject to further processing.
.IP "3. delimiter mapping" 4
For each string picked up by the last step, delimiter mapping is performed.
That is, substibute certain characters specified as domain name delimiter
with period.
.IP "4. local mapping" 4
Perform local mapping.
If the local mapping is specified by command line option ``\-localmap'',

bind/contrib/mdnkit-2.2.3-src/tools/mdnconv/mdnconv.c
#ifndef lint
static char *rcsid = "$Id: mdnconv.c,v 1.1.1.1 2002/11/18 22:25:28 bbraun Exp $";
#endif

/*
 * Copyright (c) 2000,2001 Japan Network Information Center.
 * All rights reserved.
 *  
 * By using this file, you agree to the terms and conditions set forth bellow.
 * 
 * 			LICENSE TERMS AND CONDITIONS 
 * 
 * The following License Terms and Conditions apply, unless a different
 * license is obtained from Japan Network Information Center ("JPNIC"),
 * a Japanese association, Kokusai-Kougyou-Kanda Bldg 6F, 2-3-4 Uchi-Kanda,
 * Chiyoda-ku, Tokyo 101-0047, Japan.
 * 
 * 1. Use, Modification and Redistribution (including distribution of any
 *    modified or derived work) in source and/or binary forms is permitted
 *    under this License Terms and Conditions.
 * 
 * 2. Redistribution of source code must retain the copyright notices as they
 *    appear in each source code file, this License Terms and Conditions.
 * 
 * 3. Redistribution in binary form must reproduce the Copyright Notice,
 *    this License Terms and Conditions, in the documentation and/or other
 *    materials provided with the distribution.  For the purposes of binary
 *    distribution the "Copyright Notice" refers to the following language:
 *    "Copyright (c) Japan Network Information Center.  All rights reserved."
 * 
 * 4. Neither the name of JPNIC may be used to endorse or promote products
 *    derived from this Software without specific prior written approval of
 *    JPNIC.
 * 
 * 5. Disclaimer/Limitation of Liability: THIS SOFTWARE IS PROVIDED BY JPNIC
 *    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 *    PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL JPNIC BE LIABLE
 *    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 *    BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 *    WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 *    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 *    ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * 6. Indemnification by Licensee
 *    Any person or entities using and/or redistributing this Software under
 *    this License Terms and Conditions shall defend indemnify and hold
 *    harmless JPNIC from and against any and all judgements damages,
 *    expenses, settlement liabilities, cost and other liabilities of any
 *    kind as a result of use and redistribution of this Software or any
 *    claim, suite, action, litigation or proceeding by any third party
 *    arising out of or relates to this License Terms and Conditions.
 * 
 * 7. Governing Law, Jurisdiction and Venue
 *    This License Terms and Conditions shall be governed by and and
 *    construed in accordance with the law of Japan. Any person or entities
 *    using and/or redistributing this Software under this License Terms and
 *    Conditions hereby agrees and consent to the personal and exclusive
 *    jurisdiction and venue of Tokyo District Court of Japan.
 */

/*
 * mdnconv -- Codeset converter for named.conf and zone files
 */

#include <config.h>

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif

#include <mdn/result.h>
#include <mdn/converter.h>
#include <mdn/normalizer.h>
#include <mdn/utf8.h>
#include <mdn/resconf.h>
#include <mdn/res.h>
#include <mdn/util.h>
#include <mdn/version.h>

#include "util.h"

#define MAX_DELIMITER		10
#define MAX_LOCALMAPPER		10
#define MAX_MAPPER		10
#define MAX_NORMALIZER		10
#define MAX_CHEKER		10

#define FLAG_REVERSE		1
#define FLAG_DELIMITERMAP	2
#define FLAG_LOCALMAP		4
#define FLAG_NAMEPREP		8
#define FLAG_UNASSIGNCHECK	16
#define FLAG_SELECTIVE		32

int		line_number;		/* current input file line number */
static int	flush_every_line = 0;	/* pretty obvious */

static int		convert_file(mdn_resconf_t conf, FILE *fp, int flags);
static void		print_usage(char *cmd);
static void		print_version(void);
static unsigned long	get_ucs(const char *p);

int
main(int ac, char **av) {
	char *cmd = *av;
	char *cname;
	unsigned long delimiters[MAX_DELIMITER];
	char *localmappers[MAX_LOCALMAPPER];
	char *mappers[MAX_MAPPER];
	char *normalizers[MAX_NORMALIZER];
	char *prohibits[MAX_CHEKER];
	char *unassigns[MAX_CHEKER];
	char *nameprep_version = NULL;
	int ndelimiters = 0;
	int nlocalmappers = 0;
	int nmappers = 0;
	int nnormalizers = 0;
	int nprohibits = 0;
	int nunassigns = 0;
	char *in_code = NULL;
	char *out_code = NULL;
	char *resconf_file = NULL;
	int no_resconf = 0;
	char *encoding_alias = NULL;
	int check_unassigned = 0;
	int flags = FLAG_LOCALMAP | FLAG_NAMEPREP | FLAG_SELECTIVE;
	FILE *fp;
	mdn_result_t r;
	mdn_resconf_t resconf;

#ifdef HAVE_SETLOCALE
	(void)setlocale(LC_ALL, "");
#endif

	/*
	 * If the command name begins with 'r', reverse mode is assumed.
	 */
	if ((cname = strrchr(cmd, '/')) != NULL)
		cname++;
	else
		cname = cmd;
	if (cname[0] == 'r')
		flags |= FLAG_REVERSE;

	ac--;
	av++;
	while (ac > 0 && **av == '-') {

#define OPT_MATCH(opt) (strcmp(*av, opt) == 0)
#define MUST_HAVE_ARG if (ac < 2) print_usage(cmd)
#define APPEND_LIST(array, size, item, what) \
	if (size >= (sizeof(array) / sizeof(array[0]))) { \
		errormsg("too many " what "\n"); \
		exit(1); \
	} \
	array[size++] = item; \
	ac--; av++

		if (OPT_MATCH("-in") || OPT_MATCH("-i")) {
			MUST_HAVE_ARG;
			in_code = av[1];
			ac--;
			av++;
		} else if (OPT_MATCH("-out") || OPT_MATCH("-o")) {
			MUST_HAVE_ARG;
			out_code = av[1];
			ac--;
			av++;
		} else if (OPT_MATCH("-conf") || OPT_MATCH("-c")) {
			MUST_HAVE_ARG;
			resconf_file = av[1];
			ac--;
			av++;
		} else if (OPT_MATCH("-nameprep") || OPT_MATCH("-n")) {
			MUST_HAVE_ARG;
			nameprep_version = av[1];
			ac--;
			av++;
		} else if (OPT_MATCH("-noconf") || OPT_MATCH("-C")) {
			no_resconf = 1;
		} else if (OPT_MATCH("-reverse") || OPT_MATCH("-r")) {
			flags |= FLAG_REVERSE;
		} else if (OPT_MATCH("-nolocalmap") || OPT_MATCH("-L")) {
			flags &= ~FLAG_LOCALMAP;
		} else if (OPT_MATCH("-delimitermap") || OPT_MATCH("-d")) {
			flags |= FLAG_DELIMITERMAP;
		} else if (OPT_MATCH("-nonameprep") || OPT_MATCH("-N")) {
			flags &= ~FLAG_NAMEPREP;
		} else if (OPT_MATCH("-unassigncheck") || OPT_MATCH("-u")) {
			flags |= FLAG_UNASSIGNCHECK;
		} else if (OPT_MATCH("-whole") || OPT_MATCH("-w")) {
			flags &= ~FLAG_SELECTIVE;
		} else if (OPT_MATCH("-localmap")) {
			MUST_HAVE_ARG;
			APPEND_LIST(localmappers, nlocalmappers, av[1],
				    "local maps");
		} else if (OPT_MATCH("-delimiter")) {
			unsigned long v;
			MUST_HAVE_ARG;
			v = get_ucs(av[1]);
			APPEND_LIST(delimiters, ndelimiters, v,
				    "delimiter maps");
		} else if (OPT_MATCH("-map")) {
			MUST_HAVE_ARG;
			APPEND_LIST(mappers, nmappers, av[1], "mappers");
		} else if (OPT_MATCH("-normalize")) {
			MUST_HAVE_ARG;
			APPEND_LIST(normalizers, nnormalizers, av[1],
				    "normalizers");
		} else if (OPT_MATCH("-prohibit")) {
			MUST_HAVE_ARG;
			APPEND_LIST(prohibits, nprohibits, av[1],
				    "prohibited checkers");
		} else if (OPT_MATCH("-unassigned")) {
			MUST_HAVE_ARG;
			APPEND_LIST(unassigns, nunassigns, av[1],
				    "unassigned checkers");
			check_unassigned = 1;
		} else if (OPT_MATCH("-alias") || OPT_MATCH("-a")) {
			MUST_HAVE_ARG;
			encoding_alias = av[1];
			ac--;
			av++;
		} else if (OPT_MATCH("-flush")) {
			flush_every_line = 1;
		} else if (OPT_MATCH("-version") || OPT_MATCH("-v")) {
			print_version();
		} else {
			print_usage(cmd);
		}
#undef OPT_MATCH
#undef MUST_HAVE_ARG
#undef APPEND_LIST

		ac--;
		av++;
	}

	if (ac > 1)
		print_usage(cmd);

	/* Initialize. */
	if ((r = mdn_resconf_initialize()) != mdn_success) {
		errormsg("error initializing library\n");
		return (1);
	}

	/* Create resource context. */
	resconf = NULL;
	if ((r = mdn_resconf_create(&resconf)) != mdn_success) {
		errormsg("error initializing configuration parameters\n");
		return (1);
	}

	/* Load configuration file. */
	if (!no_resconf) {
		r = mdn_resconf_loadfile(resconf, resconf_file);
		if (r != mdn_success) {
			errormsg("error reading configuration file: %s\n",
				 mdn_result_tostring(r));
			return (1);
		}
	}

	/* Set encoding alias file. */
	if (encoding_alias != NULL)
		set_encoding_alias(encoding_alias);

	/* Set input/output codeset. */
	if (flags & FLAG_REVERSE) {
		if (in_code != NULL)
			set_idncode(resconf, in_code);
		else
			check_defaultidncode(resconf, "-in");
		if (out_code != NULL)
			set_localcode(resconf, out_code);
		else
			check_defaultlocalcode(resconf, "-out");
	} else {
		if (in_code != NULL)
			set_localcode(resconf, in_code);
		else
			check_defaultlocalcode(resconf, "-in");
		if (out_code != NULL)
			set_idncode(resconf, out_code);
		else
			check_defaultidncode(resconf, "-out");
	}

	/* Set delimiter map(s). */
	if (ndelimiters > 0)
		set_delimitermapper(resconf, delimiters, ndelimiters);

	/* Set local map(s). */
	if (nlocalmappers > 0)
		set_localmapper(resconf, localmappers, nlocalmappers);

	/* Set NAMEPREP version. */
	if (nameprep_version != NULL)
		set_nameprep(resconf, nameprep_version);

	/* Set NAMEPREP mapper. */
	if (nmappers > 0)
		set_mapper(resconf, mappers, nmappers);

	/* Set NAMEPREP normalizer. */
	if (nnormalizers > 0)
		set_normalizer(resconf, normalizers, nnormalizers);

	/* Set NAMEPREP prohibit checker. */
	if (nprohibits > 0)
		set_prohibit_checkers(resconf, prohibits, nprohibits);

	/* Set NAMEPREP unassigned checker. */
	if (nunassigns > 0)
		set_unassigned_checkers(resconf, unassigns, nunassigns);

	/* Open input file. */
	if (ac > 0) {
		if ((fp = fopen(av[0], "r")) == NULL) {
			errormsg("cannot open file %s: %s\n",
				 av[0], strerror(errno));
			return (1);
		}
	} else {
		fp = stdin;
	}

	/* Do the conversion. */
	return convert_file(resconf, fp, flags);
}

static int
convert_file(mdn_resconf_t conf, FILE *fp, int flags) {
	mdn_result_t r;
	char line1[1024];
	char line2[1024];
	char insn1[10], insn2[10];
	int nl_trimmed;
	int ace_hack;
	mdn_converter_t conv;

	/*
	 * See if the input codeset is an ACE.
	 */
	if (flags & FLAG_REVERSE)
		conv = mdn_resconf_getidnconverter(conf);
	else
		conv = mdn_resconf_getlocalconverter(conf);
	if (conv != NULL && mdn_converter_isasciicompatible(conv))
		ace_hack = 1;
	else
		ace_hack = 0;
	if (conv != NULL)
		mdn_converter_destroy(conv);

	if (flags & FLAG_REVERSE) {
		char *insnp = insn1;

		*insnp++ = 'i';
		if (flags & FLAG_NAMEPREP) {
			*insnp++ = '!';
			*insnp++ = 'N';
		}
		if (flags & FLAG_UNASSIGNCHECK) {
			*insnp++ = '!';
			*insnp++ = 'u';
		}
		*insnp = '\0';

		strcpy(insn2, "L");
	} else {
		char *insnp = insn2;

		strcpy(insn1, "l");

		if (flags & FLAG_DELIMITERMAP)
			*insnp++ = 'd';
		if (flags & FLAG_LOCALMAP)
			*insnp++ = 'M';
		if (flags & FLAG_NAMEPREP)
			*insnp++ = 'N';
		if (flags & FLAG_UNASSIGNCHECK)
			*insnp++ = 'u';
		*insnp++ = 'I';
		*insnp = '\0';
	}

	line_number = 1;
	while (fgets(line1, sizeof(line1), fp) != NULL) {
		/*
		 * Trim newline at the end.  This is needed for
		 * those ascii-comatible encodings such as UTF-5 or RACE
		 * not to try converting newlines, which will result
		 * in `invalid encoding' error.
		 */
		if (line1[strlen(line1) - 1] == '\n') {
			line1[strlen(line1) - 1] = '\0';
			nl_trimmed = 1;
		} else {
			nl_trimmed = 0;
		}

		/*
		 * Convert input line to UTF-8.
		 */
		if (ace_hack && (flags & FLAG_SELECTIVE)) {
			/*
			 * Selectively decode those portions.
			 */
			r = selective_decode(conf, insn1, line1, line2,
					     sizeof(line2));
		} else {
			r = mdn_res_nameconv(conf, insn1,
					     line1, line2, sizeof(line2));
		}
		if (r != mdn_success) {
			errormsg("conversion failed at line %d: %s\n",
				 line_number,
				 mdn_result_tostring(r));
			return (1);
		}
		if (!mdn_utf8_isvalidstring(line2)) {
			errormsg("conversion to utf-8 failed at line %d\n",
				 line_number);
			return (1);
		}

		/*
		 * Perform local mapping and NAMEPREP, and convert to
		 * the output codeset.
		 */
		if (!(flags & FLAG_REVERSE) && (flags & FLAG_SELECTIVE)) {
			r = selective_encode(conf, insn2, line2, line1,
					     sizeof(line1));
		} else {
			r = mdn_res_nameconv(conf, insn2, line2, line1,
					     sizeof(line1));
		}
		if (r != mdn_success) {
			errormsg("error in nameprep or output conversion "
				 "at line %d: %s\n",
				 line_number, mdn_result_tostring(r));
			return (1);
		}

		fputs(line1, stdout);
		if (nl_trimmed)
			putc('\n', stdout);

		if (flush_every_line)
			fflush(stdout);

		line_number++;
	}
	return (0);
}

static char *options[] = {
	"-in INPUT-CODESET   : specifies input codeset name.",
	"-i INPUT-CODESET    : synonym for -in",
	"-out OUTPUT-CODESET : specifies output codeset name.",
	"-o OUTPUT-CODESET   : synonym for -out",
	"-conf CONF-FILE     : specifies pathname of MDN configuration file.",
	"-c CONF-FILE        : synonym for -conf",
	"-noconf             : do not load MDN configuration file.",
	"-C                  : synonym for -noconf",
	"-reverse            : specifies reverse conversion.",
	"                      (i.e. IDN encoding to local encoding)",
	"-r                  : synonym for -reverse",
	"-nameprep VERSION   : specifies version name of NAMEPREP.",
	"-n VERSION          : synonym for -nameprep",
	"-nonameprep         : do not perform NAMEPREP.",
	"-N                  : synonym for -nonameprep",
	"-localmap MAPPING   : specifies local mapping.",
	"-nolocalmap         : do not perform local mapping.",
	"-L                  : synonym for -nolocalmap",
	"-map SCHEME         : specifies mapping scheme.",
	"-normalize SCHEME   : specifies normalization scheme.",
	"-prohibit SET       : specifies set of prohibited characters.",
	"-unassigned SET     : specifies set of unassigned code points.",
	"-unassigncheck      : perform unassigned codepoint check.",
	"-u                  : synonym for -unassigncheck",
	"-delimiter U+XXXX   : specifies local delimiter code point.",
	"-delimitermap       : perform local delimiter mapping.",
	"-d                  : synonym for -delimitermap",
	"-alias alias-file   : specifies codeset alias file.",
	"-a                  : synonym for -alias",
	"-flush              : line-buffering mode.",
	"-whole              : convert the whole region instead of",
	"                      regions containing non-ascii characters.",
	"-w                  : synonym for -whole",
	"-version            : print version number, then exit.",
	"-v                  : synonym for -version",
	"",
	" The following options can be specified multiple times",
	"   -localmap, -map, -normalize, -prohibit, -unassigned -delimiter",
	NULL,
};

static void
print_version() {
	fprintf(stderr, "mdnconv (mDNkit) version: %s\n"
		"library version: %s\n",
		MDNKIT_VERSION,
		mdn_version_getstring());
	exit(0);
}

static void
print_usage(char *cmd) {
	int i;

	fprintf(stderr, "Usage: %s [options..] [file]\n", cmd);

	for (i = 0; options[i] != NULL; i++)
		fprintf(stderr, "\t%s\n", options[i]);

	exit(1);
}

static unsigned long
get_ucs(const char *p) {
	unsigned long v;
	char *end;

	/* Skip optional 'U+' */
	if (strncmp(p, "U+", 2) == 0)
		p += 2;

	v = strtoul(p, &end, 16);
	if (*end != '\0') {
		fprintf(stderr, "invalid UCS code point \"%s\"\n", p);
		exit(1);
	}

	return v;
}
